src/espidf.rs

```diff
@@ -338,32 +338,56 @@ pub enum FromEnvError {
 /// Information about a esp-idf source and tools installation.
 #[derive(Debug)]
 pub struct EspIdf {
-    /// The esp-idf repository.
-    pub repository: git::Repository,
+    /// The esp-idf source tree.
+    pub esp_idf_dir: SourceTree,
     /// The binary paths of all tools concatenated with the system `PATH` env variable.
     pub exported_path: OsString,
     /// The path to the python executable to be used by the esp-idf.
     pub venv_python: PathBuf,
     /// The version of the esp-idf or [`Err`] if it could not be detected.
     pub version: Result<EspIdfVersion>,
-    /// Whether [`EspIdf::repository`] is installed and managed by [`Installer`] and
-    /// **not** provided by the user.
+    /// Whether [`EspIdf::tree`] is a repository installed and managed by
+    /// [`Installer`] and **not** provided by the user.
     pub is_managed_espidf: bool,
 }
 
+#[derive(Debug, Clone)]
+pub enum SourceTree {
+    Git(git::Repository),
+    Plain(PathBuf),
+}
+
+impl SourceTree {
+    pub fn open(path: &Path) -> Self {
+        git::Repository::open(path)
+            .map(SourceTree::Git)
+            .unwrap_or_else(|_| SourceTree::Plain(path.to_owned()))
+    }
+
+    pub fn path(&self) -> &Path {
+        match self {
+            SourceTree::Git(repo) => repo.worktree(),
+            SourceTree::Plain(path) => path,
+        }
+    }
+}
+
 impl EspIdf {
     /// Try to detect an activated esp-idf environment.
-    pub fn try_from_env() -> Result<EspIdf, FromEnvError> {
-        // detect repo from $IDF_PATH
-        let idf_path = env::var_os(IDF_PATH_VAR).ok_or_else(|| {
-            FromEnvError::NoRepo(anyhow!("environment variable `{IDF_PATH_VAR}` not found"))
+    pub fn try_from_env(idf_path: Option<&Path>) -> Result<EspIdf, FromEnvError> {
```

Comment:
What is this `idf_path` which is passed from the outside, and why is it necessary? Perhaps we need to document it in the `\\\` now that the method has it, and given that (at least to me) its purpose is unclear?

Reply:
Also: if it turns out we need `idf_path` after all, isn't it better to just have two methods (which might delegate internally to a common private method still so as to avoid code duplication): 
- `EspIdf::from_paths(idf_path: &Path, path: &Path) -> Self`
- `EspIdf::from_env() -> Result<Self, FromEnvError>`

Reply:
Definitely agree re: documentation and splitting the method.

The parameter exists to make the API more sensible for the consuming build tool.

There is an existing use case which may be broken today: projects that use `esp-idf-sys` are [supposed to be able to](https://github.com/esp-rs/esp-idf-sys/blob/master/BUILD-OPTIONS.md#esp-idf-configuration) set this variable in `package.metadata.esp-idf-sys` in `Cargo.toml`, but I don't believe they can, as this can only be configured with the `IDF_PATH` env var.